### DISCUSSION TITLE
Using tree mutability across reparsing

### DISCUSSION THREAD

**Discussion by vigoux** [tree-sitter/tree-sitter#901] [Q&A] [answered]
Hi, while working (on nvim tree-sitter) on some incremental definition / completion engine using tree-sitter, I suddenly had a question.

The thing I would like to exploit is mutability of the trees (and nodes) across reparse.
So here is my question :

Le'ts say I am *re*parsing a document, with only a part of the document has changed, am I sure that all nodes that our outside the ranges returned by `ts_tree_get_changed_ranges` are the same as in the old tree?

Another question is also, can we use `node.id` reliably to refer to a node, and the "version" of a node in a tree, across reparsing ?

**Comment by maxbrunsfeld** [comment, marked as answer]
> am I sure that all nodes that our outside the ranges returned by ts_tree_get_changed_ranges are the same as in the old tree?

They should be identical, yes. But there be some nodes that have been recreated (instead of being reused).

> can we use node.id reliably to refer to a node, and the "version" of a node in a tree, across reparsing ?

If two nodes' `id` values are the same, then the nodes share a pointer representation, so their internal contents are guaranteed to be the same. However, if node's `id` values are different, they may *still* be the same. The node-reuse algorithm is sometimes conservative, and nodes may be recreated even if their contents haven't changed.

**Reply by vigoux** [reply]
Great, thanks for your answer!
